Move API for file mapping to the new FileHandle class

cdumez opened the pull request:
#### d8d038849797ffa70e6614b51dba9e6ca42b7c6c
<pre>
Move API for file mapping to the new FileHandle class
<a href="https://bugs.webkit.org/show_bug.cgi?id=290022">https://bugs.webkit.org/show_bug.cgi?id=290022</a>

Reviewed by Darin Adler.

Move API for file mapping to the new FileHandle class since the new main entry point
to deal with file descriptors.

It used to be constructors and create() functions on the MappedFileData class, taking
a FileHandle as parameter. It&apos;s easier to just call `map()` on the FileHandle directly
and is more consistent with how other FileSystem APIs currently work.

* Source/JavaScriptCore/API/JSScript.mm:
(+[JSScript scriptOfType:memoryMappedFromASCIIFile:withSourceURL:andBytecodeCache:inVirtualMachine:error:]):
(-[JSScript readCache]):
* Source/JavaScriptCore/jsc.cpp:
* Source/JavaScriptCore/runtime/CachedTypes.cpp:
(JSC::Encoder::releaseMapped):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/FileHandle.h:
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::mapFile):
(WTF::FileSystemImpl::createMappedFileData):
(WTF::FileSystemImpl::MappedFileData::MappedFileData): Deleted.
(WTF::FileSystemImpl::MappedFileData::mapFileHandle): Deleted.
* Source/WTF/wtf/FileSystem.h:
(WTF::FileSystemImpl::MappedFileData::leakHandle): Deleted.
(WTF::FileSystemImpl::MappedFileData::operator bool const): Deleted.
(WTF::FileSystemImpl::MappedFileData::size const): Deleted.
(): Deleted.
(WTF::FileSystemImpl::MappedFileData::fileMapping const): Deleted.
(WTF::FileSystemImpl::MappedFileData::span const): Deleted.
(WTF::FileSystemImpl::MappedFileData::mutableSpan): Deleted.
(WTF::FileSystemImpl::MappedFileData::create): Deleted.
(WTF::FileSystemImpl::MappedFileData::MappedFileData): Deleted.
(WTF::FileSystemImpl::m_fileMapping): Deleted.
(WTF::FileSystemImpl::MappedFileData::operator=): Deleted.
* Source/WTF/wtf/MappedFileData.h: Added.
(WTF::FileSystemImpl::MappedFileData::leakHandle):
(WTF::FileSystemImpl::MappedFileData::operator bool const):
(WTF::FileSystemImpl::MappedFileData::size const):
(WTF::FileSystemImpl::MappedFileData::fileMapping const):
(WTF::FileSystemImpl::MappedFileData::span const):
(WTF::FileSystemImpl::MappedFileData::mutableSpan):
* Source/WTF/wtf/PlatformJSCOnly.cmake:
* Source/WTF/wtf/PlatformWin.cmake:
* Source/WTF/wtf/posix/FileHandlePOSIX.cpp:
(WTF::FileSystemImpl::FileHandle::lock):
(WTF::FileSystemImpl::FileHandle::map):
* Source/WTF/wtf/posix/MappedFileDataPOSIX.cpp: Added.
(WTF::FileSystemImpl::MappedFileData::MappedFileData):
* Source/WTF/wtf/win/FileHandleWin.cpp:
(WTF::FileSystemImpl::FileHandle::map):
* Source/WTF/wtf/win/FileSystemWin.cpp:
(WTF::FileSystemImpl::MappedFileData::~MappedFileData): Deleted.
(WTF::FileSystemImpl::MappedFileData::mapFileHandle): Deleted.
* Source/WTF/wtf/win/MappedFileDataWin.cpp: Added.
(WTF::FileSystemImpl::MappedFileData):
(WTF::FileSystemImpl::MappedFileData::~MappedFileData):
* Source/WebCore/contentextensions/SerializedNFA.cpp:
(WebCore::ContentExtensions::SerializedNFA::serialize):
* Source/WebCore/platform/SharedBuffer.cpp:
(WebCore::SharedBuffer::createWithContentsOfFile):
* Source/WebKit/NetworkProcess/cache/NetworkCacheData.cpp:
(WebKit::NetworkCache::adoptAndMapFile):
* Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
(WebKit::CacheStorageDiskStore::readAllRecordInfosInternal):
(WebKit::CacheStorageDiskStore::readRecordsInternal):
* Tools/TestWebKitAPI/Tests/WTF/FileSystem.cpp:
(TestWebKitAPI::TEST_F(FileSystemTest, MappingMissingFile)):
(TestWebKitAPI::TEST_F(FileSystemTest, MappingExistingFile)):
(TestWebKitAPI::TEST_F(FileSystemTest, MappingExistingEmptyFile)):

Canonical link: <a href="https://commits.webkit.org/292461@main">https://commits.webkit.org/292461@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/814efcdff23e17295ec3576450a4fc26f529ee8d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96135 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15749 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5707 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101200 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46644 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98180 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16044 "Build is in progress. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Running apply-patch; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24182 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73287 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/30508 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99138 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12017 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86839 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53624 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11767 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4600 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45979 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/88808 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/81901 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4697 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103225 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/94756 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23202 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/16894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/82326 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23453 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82860 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81702 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20496 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26311 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/3745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16558 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23165 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/28320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/118233 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22824 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26304 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24565 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/API/JSScript.mm:
```diff
@@ -194,7 +192,7 @@ - (void)readCache
     }
 
-    Ref<JSC::CachedBytecode> cachedBytecode = JSC::CachedBytecode::create(WTFMove(mappedFile));
+    Ref<JSC::CachedBytecode> cachedBytecode = JSC::CachedBytecode::create(WTFMove(*mappedFile));
```
```suggestion
    Ref cachedBytecode = JSC::CachedBytecode::create(WTFMove(*mappedFile));
```

darinadler commented on Source/WTF/wtf/FileHandle.h:
```diff
@@ -27,6 +27,7 @@
 
 #include <optional>
 #include <wtf/Forward.h>
+#include <wtf/MappedFileData.h>
```
Seems a bit of a shame that we include this rather than forward declaring. Would be nicer if only callers that are mapping file data needed to include this header.

cdumez replied:
I will fix that before landing.

cdumez committed:
Move API for file mapping to the new FileHandle class
https://bugs.webkit.org/show_bug.cgi?id=290022

Reviewed by Darin Adler.

Move API for file mapping to the new FileHandle class since the new main entry point
to deal with file descriptors.

It used to be constructors and create() functions on the MappedFileData class, taking
a FileHandle as parameter. It's easier to just call `map()` on the FileHandle directly
and is more consistent with how other FileSystem APIs currently work.

* Source/JavaScriptCore/API/JSScript.mm:
(+[JSScript scriptOfType:memoryMappedFromASCIIFile:withSourceURL:andBytecodeCache:inVirtualMachine:error:]):
(-[JSScript readCache]):
* Source/JavaScriptCore/jsc.cpp:
* Source/JavaScriptCore/runtime/CachedTypes.cpp:
(JSC::Encoder::releaseMapped):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/FileHandle.h:
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::mapFile):
(WTF::FileSystemImpl::createMappedFileData):
(WTF::FileSystemImpl::MappedFileData::MappedFileData): Deleted.
(WTF::FileSystemImpl::MappedFileData::mapFileHandle): Deleted.
* Source/WTF/wtf/FileSystem.h:
(WTF::FileSystemImpl::MappedFileData::leakHandle): Deleted.
(WTF::FileSystemImpl::MappedFileData::operator bool const): Deleted.
(WTF::FileSystemImpl::MappedFileData::size const): Deleted.
(): Deleted.
(WTF::FileSystemImpl::MappedFileData::fileMapping const): Deleted.
(WTF::FileSystemImpl::MappedFileData::span const): Deleted.
(WTF::FileSystemImpl::MappedFileData::mutableSpan): Deleted.
(WTF::FileSystemImpl::MappedFileData::create): Deleted.
(WTF::FileSystemImpl::MappedFileData::MappedFileData): Deleted.
(WTF::FileSystemImpl::m_fileMapping): Deleted.
(WTF::FileSystemImpl::MappedFileData::operator=): Deleted.
* Source/WTF/wtf/MappedFileData.h: Added.
(WTF::FileSystemImpl::MappedFileData::leakHandle):
(WTF::FileSystemImpl::MappedFileData::operator bool const):
(WTF::FileSystemImpl::MappedFileData::size const):
(WTF::FileSystemImpl::MappedFileData::fileMapping const):
(WTF::FileSystemImpl::MappedFileData::span const):
(WTF::FileSystemImpl::MappedFileData::mutableSpan):
* Source/WTF/wtf/PlatformJSCOnly.cmake:
* Source/WTF/wtf/PlatformWin.cmake:
* Source/WTF/wtf/posix/FileHandlePOSIX.cpp:
(WTF::FileSystemImpl::FileHandle::lock):
(WTF::FileSystemImpl::FileHandle::map):
* Source/WTF/wtf/posix/MappedFileDataPOSIX.cpp: Added.
(WTF::FileSystemImpl::MappedFileData::MappedFileData):
* Source/WTF/wtf/win/FileHandleWin.cpp:
(WTF::FileSystemImpl::FileHandle::map):
* Source/WTF/wtf/win/FileSystemWin.cpp:
(WTF::FileSystemImpl::MappedFileData::~MappedFileData): Deleted.
(WTF::FileSystemImpl::MappedFileData::mapFileHandle): Deleted.
* Source/WTF/wtf/win/MappedFileDataWin.cpp: Added.
(WTF::FileSystemImpl::MappedFileData):
(WTF::FileSystemImpl::MappedFileData::~MappedFileData):
* Source/WebCore/contentextensions/SerializedNFA.cpp:
(WebCore::ContentExtensions::SerializedNFA::serialize):
* Source/WebCore/platform/SharedBuffer.cpp:
(WebCore::SharedBuffer::createWithContentsOfFile):
* Source/WebKit/NetworkProcess/cache/NetworkCacheData.cpp:
(WebKit::NetworkCache::adoptAndMapFile):
* Source/WebKit/NetworkProcess/storage/CacheStorageDiskStore.cpp:
(WebKit::CacheStorageDiskStore::readAllRecordInfosInternal):
(WebKit::CacheStorageDiskStore::readRecordsInternal):
* Tools/TestWebKitAPI/Tests/WTF/FileSystem.cpp:
(TestWebKitAPI::TEST_F(FileSystemTest, MappingMissingFile)):
(TestWebKitAPI::TEST_F(FileSystemTest, MappingExistingFile)):
(TestWebKitAPI::TEST_F(FileSystemTest, MappingExistingEmptyFile)):

Canonical link: https://commits.webkit.org/292461@main
